src/cljc/andel/editor.cljc

```diff
@@ -0,0 +1,146 @@
+(ns andel.editor
+  (:require [andel.text :as text]
+            [andel.utils :as utils]
+            [andel.controller :as contr]))
+
+(defn backspace [{:keys [document editor] :as state}]
+  (let [{:keys [text]} document
+        {:keys [caret selection]} editor
+        {caret-offset :offset} caret
+        [sel-from sel-to] selection
+        sel-len (- sel-to sel-from)]
+    (cond (< 0 sel-len)
+          (contr/delete-under-selection state)
+
+          (< 0 caret-offset)
+          (-> state
+              (contr/edit-at-offset (dec caret-offset) #(text/delete % 1))
+              (assoc-in [:editor :caret :offset] (dec caret-offset))
+              (assoc-in [:editor :caret :v-col] 0)
+              (assoc-in [:editor :selection] [(dec caret-offset) (dec caret-offset)]))
```

Comment:
I would argue that using zero-len selection is better at avoiding edge cases. If you have something to do with selection you should consider 2 different possibilities. Empty vector is way worse because then you get nils after destructuring and NPEs.